src/Input/AutoResizingTextArea.js

```diff
@@ -0,0 +1,204 @@
+// @flow weak
+
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { createStyleSheet } from 'jss-theme-reactor';
+import classnames from 'classnames';
+import EventListener from 'react-event-listener';
+import customPropTypes from '../utils/customPropTypes';
+
+const rowsHeight = 24;
+
+export const styleSheet = createStyleSheet('MuiTextarea', () => {
+  return {
+    root: {
+      position: 'relative', // because the shadow has position: 'absolute',
+      'margin-bottom': '-4px', // this is an unfortunate hack
+    },
+    textarea: {
+      width: '100%',
+      resize: 'none',
+      font: 'inherit',
+      padding: 0,
+      cursor: 'inherit',
+      boxSizing: 'border-box',
+      lineHeight: 'inherit',
+      border: 'none',
+      outline: 'none',
+      'background-color': 'rgba(0,0,0,0)',
+    },
+    shadow: {
+      resize: 'none',
+      // Overflow also needed to here to remove the extra row
+      // added to textareas in Firefox.
+      overflow: 'hidden',
+      // Visibility needed to hide the extra text area on ipads
+      visibility: 'hidden',
+      position: 'absolute',
+      height: 'auto',
+      whiteSpace: 'pre-wrap',
+    },
+  };
+});
+
+/**
+ * Input
+ */
+export default class AutoResizingTextArea extends Component {
+  shadow: HTMLInputElement;
+  singleLineShadow: HTMLInputElement;
+  input: HTMLInputElement;
+  value: string;
+
+  static propTypes = {
+    /**
+     * Override the inline-styles of the root element.
+     */
+    className: PropTypes.string,
+    defaultValue: PropTypes.any,
+    disabled: PropTypes.bool,
+    hintText: PropTypes.string,
+    onChange: PropTypes.func,
+    onHeightChange: PropTypes.func,
+    rows: PropTypes.number,
+    rowsMax: PropTypes.number,
+    shadowClassName: PropTypes.object,
+    value: PropTypes.string,
+  };
+
+  static defaultProps = {
+    rows: 1,
+  };
+
+  static contextTypes = {
+    styleManager: customPropTypes.muiRequired,
+  };
+
+  state = {
+    height: null,
+  };
+
+  componentWillMount() {
+    // <Input> expects the components it renders to respond to 'value'
+    this.value = this.props.defaultValue;
+    this.setState({
+      height: this.props.rows * rowsHeight,
+    });
+  }
+
+  componentDidMount() {
+    this.syncHeightWithShadow();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.value !== this.props.value ||
+        nextProps.rowsMax !== this.props.rowsMax) {
+      this.syncHeightWithShadow(nextProps.value, null, nextProps);
+    }
+  }
+
+  handleResize = (event) => {
```

Comment:
I think that we should debounce the events, for instance:
https://github.com/callemall/material-ui/blob/next/src/Tabs/Tabs.js#L163-L173